mlrun/datastore/base.py

```diff
@@ -157,6 +158,178 @@ def upload(self, key, src_path):
+
+        if not timing and directory_end.endswith((".parquet", ".pq")):
+            paths.append(directory_start.rstrip("/"))
+            return
+        elif timing and timing == partition_level:
+            paths.append(current_path.rstrip("/"))
+            return
+
+        directories = filesystem.ls(current_path, detail=True)
+        if len(directories) == 0:
+            return
+        for directory in directories:
+            current_path = directory["name"]
+            parts = [p for p in current_path.split("/") if "=" in p]
+            kwargs = {}
+            for part in parts:
+                part = part.split("=")
+                key = part[0]
+                value = int(part[1]) if part[1].isdigit() else part[1]
+                kwargs[key] = value
+            if DataStore._is_directory_in_range(start_time, end_time, **kwargs):
+                DataStore._list_partition_paths_helper(
+                    paths,
+                    start_time,
+                    end_time,
+                    current_path,
+                    partition_level,
+                    filesystem,
+                )
+
+    @staticmethod
+    def _list_partitioned_paths(
+        base_path: str,
+        start_time: datetime.datetime,
+        end_time: datetime.datetime,
+        partition_level: str,
+        filesystem,
+    ):
+        paths = []
+
+        DataStore._list_partition_paths_helper(
+            paths, start_time, end_time, base_path, partition_level, filesystem
+        )
+        for i in range(len(paths)):
+            paths[i] = DataStore._reconstruct_path_from_base_path(base_path, paths[i])
+        return paths
+
+    @staticmethod
+    def _reconstruct_path_from_base_path(base_path: str, returned_path: str) -> str:
+        base_path = base_path.rstrip("/")
+        clean_return = returned_path.strip("/")
+
+        prefix, _, base_suffix = base_path.partition(":///")
+        schema = prefix + ":///"
+
+        return schema + clean_return
+
+    @staticmethod
+    def _clean_filters_for_partitions(
+        filters,
+        partition_keys,
+    ):
+        """
+        Remove partition keys from filters.
+
+        :param filters: (list of lists of tuples) pandas-style filters
+                Example: [[('year','=',2025),('month','=',11),('timestamp','>',ts1)]]
+        :param partition_keys: (list of str) partition columns handled via directory
+
+        :return list of list of tuples: cleaned filters
+        """
+        cleaned_filters = []
+        for group in filters:
+            new_group = [f for f in group if f[0] not in partition_keys]
+            if new_group:
+                cleaned_filters.append(new_group)
+        return cleaned_filters
+
+    @staticmethod
+    def _read_partitioned_parquet(
+        base_path: str,
+        start_time: datetime.datetime,
+        end_time: datetime.datetime,
+        partition_keys: list[str],
+        df_module: Union[pd, object],
```

Comment:
`Union[pd, object]`? Pretty sure everything is an `object` in Python, so this is equivalent to just `object`.